### PR TITLE
C2c call support

### PR DESCRIPTION
**Pull request by vuvth**
## Proposed Changes
+ Support c2c call 
+ Only allow CallApp action(not support DeployApp, UpdateApp... for now).

## Notes 
The main point need to note on my PR is if one appl inner transaction failed. The context(ctx) will not apply any changed 
It's mean if appl A->B->C. and transaction B->C failed. we will roll back to the state before we call.

### TODO:
- [x] Fix can't use TypeNum for teal v6(keyreg and appl) 
- [x] Implement and add missing test for  `ITxna` opcode.
- [x] Inner tx only can call to app implement in teal v6
- [x] Implement and test call app 
- [x] NoOp with other appl type transaction. Only allow NoOpt(AppCall) 
- [x] Update documents and changelog

### CODE REVIEW

File: packages/runtime/src/ctx.ts [hunk 1]
```diff
@@ -412,15 +415,34 @@ export class Ctx implements Context {
...
+	verifyAndUpdateInnerApplCallStack(): void {
+		// verify
+		if (!this.isInnerTx) return;
+
+		if (this.innerTxAppIDCallStack.length >= 8) {
+			throw new RuntimeError(RUNTIME_ERRORS.TRANSACTION.INNER_APPL_DEEP_EXCEEDED);
+		}
+		const appID = this.tx.apid ?? 0;
+		if (appID > 0 && this.innerTxAppIDCallStack.find((id) => id === appID) !== undefined) {
+			throw new RuntimeError(RUNTIME_ERRORS.TRANSACTION.INNER_APPL_SELF_CALL);
```
Comment (sczembor) [inline packages/runtime/src/ctx.ts:439]:
```suggestion
			throw new RuntimeError(RUNTIME_ERRORS.TRANSACTION.INNER_APP_SELF_CALL);
```

File: packages/runtime/src/ctx.ts [hunk 2]
```diff
@@ -662,6 +684,7 @@ export class Ctx implements Context {
 		let r: TxReceipt;
 
 		this.verifyMinimumFees();
+		this.verifyAndUpdateInnerApplCallStack();
```
Comment (sczembor) [inline packages/runtime/src/ctx.ts:687]:
```suggestion
		this.verifyAndUpdateInnerAppCallStack();
```

File: packages/runtime/src/errors/errors-list.ts [hunk 3]
```diff
@@ -533,17 +527,41 @@ const transactionErrors = {
...
+		title: "Transaction cannot close account to its sender.",
+		description: "Transaction cannot close account to its sender",
+	},
+	INNER_APP_CALL_INVALID_VERSION: {
+		number: 1408,
+		message: "Inner app call in older version %tealVersion%",
+		title: "Inner app call with version %tealVersion%.",
+		description: "Inner app call with version %tealVersion%",
+	},
+	INNER_APPL_DEEP_EXCEEDED: {
```
Comment (sczembor) [inline packages/runtime/src/errors/errors-list.ts:553]:
```suggestion
	INNER_APP_DEEP_EXCEEDED: {
```

File: packages/runtime/src/errors/errors-list.ts [hunk 4]
```diff
@@ -533,17 +527,41 @@ const transactionErrors = {
...
+		number: 1409,
+		message: "Inner transaction appl deep exceeded",
+		title: "Inner transaction appl deep exceeded.",
+		description: "Inner transaction appl deep exceeded",
+	},
+	INNER_APPL_SELF_CALL: {
```
Comment (sczembor) [inline packages/runtime/src/errors/errors-list.ts:559]:
```suggestion
	INNER_APP_SELF_CALL: {
```

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 5]
```diff
@@ -4106,6 +4110,20 @@ export class ITxnSubmit extends Op {
 		const contractAddress = getApplicationAddress(appID);
 		const contractAccount = this.interpreter.runtime.getAccount(contractAddress).account;
 
+		// only support call app(NoOpt) for appl transaction type.
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:4113]:
```suggestion
		// Supports only calling app(NoOpt) for app transaction type.
```

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 6]
```diff
@@ -4106,6 +4110,20 @@ export class ITxnSubmit extends Op {
...
+		if (
+			this.interpreter.subTxn.type === TransactionTypeEnum.APPLICATION_CALL &&
+			!isEncTxApplicationCall(this.interpreter.subTxn)
+		) {
+			console.warn("Only support application call in this version");
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:4118]:
```suggestion
			console.warn("Only supports application call in this version");
```

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 7]
```diff
@@ -4106,6 +4110,20 @@ export class ITxnSubmit extends Op {
...
+			return;
+		}
+
+		// this fee will use for pay in application call next time
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:4122]:
```suggestion
		// this fee will be used to pay for application call next time
```

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 8]
```diff
@@ -4073,24 +4066,35 @@ export class ITxnSubmit extends Op {
...
 		totalFee += this.interpreter.subTxn.fee ?? 0;
-		const totalTxCnt =
-			this.interpreter.runtime.ctx.gtxs.length + this.interpreter.innerTxns.length + 1;
+
+		let totalTxCnt = this.interpreter.innerTxns.length + 1;
+		if (!this.interpreter.runtime.ctx.isInnerTx) {
+			totalTxCnt += this.interpreter.runtime.ctx.gtxs.length;
+		}
 
 		// fee too less accross pool
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:4099]: this is unclear

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 9]
```diff
@@ -4117,23 +4135,34 @@ export class ITxnSubmit extends Op {
...
+			// execute innner transaction
+			this.interpreter.runtime.ctx.processTransactions([execParams]);
+			// update current txns to base (top-level) after innerTx execution
+
+			this.interpreter.runtime.ctx.tx = baseCurrTx;
+			this.interpreter.runtime.ctx.gtxs = baseCurrTxGrp;
+			// save executed tx
+			this.interpreter.innerTxns.push(this.interpreter.subTxn);
+		} catch (err: any) {
+			// revert to begining context
```
Comment (sczembor) [inline packages/runtime/src/interpreter/opcode-list.ts:4157]:
```suggestion
			// revert to initial context
```

File: packages/runtime/src/lib/itxn.ts [hunk 10]
```diff
@@ -160,61 +170,66 @@ export function setInnerTxField(
...
-				errMsg = "URL must not be longer than 96 bytes";
-			}
-			break;
-		}
+  // handle individual cases
+  let errMsg = "";
+  switch (field) {
+    case 'Type': {
+      const txType = txValue as string;
+      // check txType supported in current teal version or not
```
Comment (sczembor) [inline packages/runtime/src/lib/itxn.ts:178]:
```suggestion
      // check if txType is supported in current teal version
```

File: packages/runtime/test/src/lib/txn.ts [hunk 11]
```diff
@@ -38,6 +38,10 @@ describe("Convert encoded Txn to ExecParams", function () {
 			encTx.clearProgram = execParams.clearProgram;
 		}
 
+		// convert appArgs to buffer to easier compare.
```
Comment (sczembor) [inline packages/runtime/test/src/lib/txn.ts:41]:
```suggestion
		// convert appArgs to buffer for an easier comparison
```

File: packages/runtime/test/integration/c2c-call/c2c-call.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+			};
+
+			expectRuntimeError(
+				() => runtime.executeTx(execParams),
+				RUNTIME_ERRORS.TRANSACTION.INNER_APP_CALL_INVALID_VERSION,
+				"RUNTIME_ERR1408: Inner app call in older version 5"
+			);
+		});
+
+		it("should failed: inner tx appl self-call", () => {
```
Comment (sczembor) [inline packages/runtime/test/integration/c2c-call/c2c-call.ts:102]:
```suggestion
		it("should fail: inner tx app self-call", () => {
```

File: packages/runtime/test/integration/c2c-call/c2c-call.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+					appArgs: ["int:7", ...apps.map((info) => `int:${info.appID}`)],
+					payFlags: {
+						totalFee: 10000,
+					},
+				};
+
+				assert.doesNotThrow(() => runtime.executeTx(execParams));
+			});
+
+			it("Should failed: inner call with depth > 8", () => {
```
Comment (sczembor) [inline packages/runtime/test/integration/c2c-call/c2c-call.ts:155]:
```suggestion
			it("Should fail: inner call with depth > 8", () => {
```

**Review by sczembor — APPROVED** [review]

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 8 above)
Comment (vuvth) [reply packages/runtime/src/interpreter/opcode-list.ts:4099]: Sure! I also think how to improve it.

Comment (vuvth) [reply packages/runtime/src/interpreter/opcode-list.ts:4099]: I changed it in #626

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 14]
```diff
@@ -4073,24 +4066,35 @@ export class ITxnSubmit extends Op {
...
 	execute(_stack: TEALStack): void {
 		if (typeof this.interpreter.subTxn === "undefined") {
 			throw new RuntimeError(RUNTIME_ERRORS.TEAL.ITXN_SUBMIT_WITHOUT_ITXN_BEGIN, {
 				line: this.line,
 			});
 		}
 
+		// back up current context
+		if (this.interpreter.runtime.remainCtx === undefined) {
+			this.interpreter.runtime.remainCtx = cloneDeep(this.interpreter.runtime.ctx);
```
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/opcode-list.ts:4079]: let's rename `remainCtx` to `parentCtx`

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 15]
```diff
@@ -4073,24 +4066,35 @@ export class ITxnSubmit extends Op {
...
 			totalFee += t.fee ?? 0;
 		}
+
 		totalFee += this.interpreter.subTxn.fee ?? 0;
-		const totalTxCnt =
-			this.interpreter.runtime.ctx.gtxs.length + this.interpreter.innerTxns.length + 1;
+
+		let totalTxCnt = this.interpreter.innerTxns.length + 1;
+		if (!this.interpreter.runtime.ctx.isInnerTx) {
+			totalTxCnt += this.interpreter.runtime.ctx.gtxs.length;
```
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/opcode-list.ts:4096]: what if other transactions in the group have inner transactions ?

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 16]
```diff
@@ -4117,23 +4135,34 @@ export class ITxnSubmit extends Op {
...
-		this.interpreter.runtime.ctx.isInnerTx = false;
-		this.interpreter.innerTxns.push(this.interpreter.subTxn);
-		this.interpreter.subTxn = undefined;
+		try {
+			const baseCurrTx = cloneDeep(this.interpreter.runtime.ctx.tx);
+			const baseCurrTxGrp = cloneDeep(this.interpreter.runtime.ctx.gtxs);
+
+			// set up context for inner transaction
+			this.interpreter.runtime.ctx.tx = this.interpreter.subTxn;
+			this.interpreter.runtime.ctx.gtxs = [this.interpreter.subTxn];
```
[the remark this answers is not in the document]
Comment (vuvth) [reply packages/runtime/src/interpreter/opcode-list.ts:4145]: I will fix it after we merge `itxn_next` to dev.

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 15 above)
Comment (vuvth) [reply packages/runtime/src/interpreter/opcode-list.ts]: Added task for this case.

**Pull request merged**